### PR TITLE
Adapt System.CommandLine (v2) in more places

### PR DESCRIPTION
**Pull request by am11**
Continuation of #66929; switches ilc and crossgen2 to System.CommandLine.

In 4731fdb6830141fb94fa8f3dbf81c437daa56ba9, CG2 was switched from System.CommandLine (v1) to an internal CommandLine utility. Since then, the performance and memory consumption in System.CommandLine has improved quite a lot (in the effort to make it AoT friendly). SDK and some tools in runtime repo are already using System.CommandLine v2 beta4.

I did some quick testing in bash by running System.Private.CoreLib's crossgen'ing common (captured from clr subset build):
```sh
$ cat cmd.sh
/runtime/dotnet.sh \
  /runtime/artifacts/bin/coreclr/Linux.x64.Debug/crossgen2/crossgen2.dll \
  -o:/runtime/artifacts/bin/coreclr/Linux.x64.Debug/System.Private.CoreLib.dll \
  -r:/runtime/artifacts/bin/coreclr/Linux.x64.Debug/IL/*.dll \
  --targetarch:x64 -O --verify-type-and-field-layout \
  /runtime/artifacts/bin/coreclr/Linux.x64.Debug/IL/System.Private.CoreLib.dll \
  --perfmap-format-version:1 --perfmap --perfmap-path:/runtime/artifacts/bin/coreclr/Linux.x64.Debug/
```

then used GNU's time command to get process-level stats (10 iterations):

```sh
$ command time -v sh -c 'for i in $(seq 10); do sh cmd.sh; done'
```

Results:
<table>
<tr><th>Before</th><th> After </th></tr>
<tr align=left><th><pre>
  User time (seconds): 1382.99
  System time (seconds): 34.68
  Percent of CPU this job got: 493%
  Elapsed (wall clock) time (h:mm:ss or m:ss): 4:47.05
  Average shared text size (kbytes): 0
  Average unshared data size (kbytes): 0
  Average stack size (kbytes): 0
  Average total size (kbytes): 0
  Maximum resident set size (kbytes): 1522000
  Average resident set size (kbytes): 0
  Major (requiring I/O) page faults: 30242
  Minor (reclaiming a frame) page faults: 4960009
  Voluntary context switches: 371025
  Involuntary context switches: 324012
  Swaps: 0
  File system inputs: 5528
  File system outputs: 372160
  Socket messages sent: 0
  Socket messages received: 0
  Signals delivered: 0
  Page size (bytes): 4096
  Exit status: 0</pre></th><th>  <pre>
  User time (seconds): 1386.69
  System time (seconds): 34.94
  Percent of CPU this job got: 490%
  Elapsed (wall clock) time (h:mm:ss or m:ss): 4:49.87
  Average shared text size (kbytes): 0
  Average unshared data size (kbytes): 0
  Average stack size (kbytes): 0
  Average total size (kbytes): 0
  Maximum resident set size (kbytes): 1496544
  Average resident set size (kbytes): 0
  Major (requiring I/O) page faults: 30658
  Minor (reclaiming a frame) page faults: 5020581
  Voluntary context switches: 393065
  Involuntary context switches: 340019
  Swaps: 0
  File system inputs: 232
  File system outputs: 372160
  Socket messages sent: 0
  Socket messages received: 0
  Signals delivered: 0
  Page size (bytes): 4096
  Exit status: 0</pre>   </th></tr></table>

speed is pretty much in noise range; memory and I/O has slightly improved as we are now accessing options lazily from `ParseResult` (in most cases).

---

The last tool on internal plan is dotnet-pgo. If this PR gets accepted, I will work on that and delete the internal utility in a follow-up PR.

### CODE REVIEW

**Comment by am11** [comment]
@MichalStrehovsky, I was motivated by aot related [improvements](https://github.com/dotnet/command-line-api/pulls?q=is%3Apr+author:adamsitnik+is:closed) made in System.CommandLine. The speed overhead is becoming insignificant.

helloworld script:
```sh
$ /runtime/dotnet.sh \
  /runtime/artifacts/bin/coreclr/Linux.x64.Release/crossgen2/crossgen2.dll \
  -o:./helloworld.ni.dll \
  -r:/runtime/artifacts/bin/coreclr/Linux.x64.Release/IL/*.dll \
  --targetarch:x64 -O --verify-type-and-field-layout \
  /helloworld/bin/Release/net7.0/helloworld.dll \
  --perfmap-format-version:1 --perfmap --perfmap-path:/tmp/
```

Results from 100 iterations:
<table>
<tr><th>Before</th><th>After</th></tr>
<tr><td><pre>
  User time (seconds): 42.80
  System time (seconds): 9.66
  Percent of CPU this job got: 100%
  Elapsed (wall clock) time (h:mm:ss or m:ss): 0:52.21
  Average shared text size (kbytes): 0
  Average unshared data size (kbytes): 0
  Average stack size (kbytes): 0
  Average total size (kbytes): 0
  Maximum resident set size (kbytes): 128484
  Average resident set size (kbytes): 0
  Major (requiring I/O) page faults: 203893
  Minor (reclaiming a frame) page faults: 2864972
  Voluntary context switches: 16773
  Involuntary context switches: 3897
  Swaps: 0
  File system inputs: 4440
  File system outputs: 1616
  Socket messages sent: 0
  Socket messages received: 0
  Signals delivered: 0
  Page size (bytes): 4096
  Exit status: 0
</pre></td><td><pre>
  User time (seconds): 43.60
  System time (seconds): 10.49
  Percent of CPU this job got: 100%
  Elapsed (wall clock) time (h:mm:ss or m:ss): 0:53.99
  Average shared text size (kbytes): 0
  Average unshared data size (kbytes): 0
  Average stack size (kbytes): 0
  Average total size (kbytes): 0
  Maximum resident set size (kbytes): 131844
  Average resident set size (kbytes): 0
  Major (requiring I/O) page faults: 210921
  Minor (reclaiming a frame) page faults: 2916726
  Voluntary context switches: 16441
  Involuntary context switches: 3525
  Swaps: 0
  File system inputs: 0
  File system outputs: 1600
  Socket messages sent: 0
  Socket messages received: 0
  Signals delivered: 0
  Page size (bytes): 4096
  Exit status: 0
</pre></th></tr>
</table>

**Comment by am11** [comment]
cc @jkotas, @MichalStrehovsky, PTAL. Diffs without whitespace change will make it a bit easier to review. I switched to use file level namespace in Program.cs to decrease the overall indentation,

> ilc: --insructionset, crossgen2: --instruction-set

Should we align them?

**Comment by jkotas** [comment]
> I switched to use file level namespace in Program.cs to decrease the overall indentation,

That should be separate change.

File: src/coreclr/tools/Common/TypeSystem/Common/TargetArchitecture.cs
Comment (jkotas) [inline src/coreclr/tools/Common/TypeSystem/Common/TargetArchitecture.cs:17]: What is this used for?

This would mean that all places the check for `TargetArchitecture.ARM` have to check for `TargetArchitecture.ARMEL` too?

**Comment by jkotas** [comment]
ilc: --insructionset, crossgen2: --instruction-set

Yes, I think it makes sense to align there.

File: src/coreclr/tools/Common/TypeSystem/Common/TargetArchitecture.cs
Comment (am11) [reply src/coreclr/tools/Common/TypeSystem/Common/TargetArchitecture.cs:17]: It is used for (easier) strongly typed binding. In Program.ctor, it overrides `ARMEL` -> `ARM` since the rest of the implementation expects it with a flag to indicate if `ARMEL` ABI is to be used.

Alternatively, I can use the old approach to capture the flag during parsing (it is already using custom step in call to `Option<T>` ctor for --targetarch).

Comment (jkotas) [reply src/coreclr/tools/Common/TypeSystem/Common/TargetArchitecture.cs:17]: I think that would be better - since we do not expect this value to be valid, except in command line parsing.

**Comment by am11** [comment]
> We want to be consistent across the whole repro

Sure, I will revert it. `git grep '^namespace.*;' src/coreclr/tools/aot` on main shows that only `ReadyToRunDiagnosticsConstants.cs` is using file namespace.

> Yes, I think it makes sense to align these.

I will update ILC to use `--instruction-set`, which is typically used in multi-word arg names.

**Comment by jkotas** [comment]
@trylek @mangod9 Are you ok with taking this conversion for .NET 7 or would you prefer to wait until we fork for .NET 8?

**Comment by trylek** [comment]
No pushback from me, the perf results look reasonable, I believe we can take it for .NET 7.

File: src/coreclr/tools/Common/TypeSystem/Common/TargetArchitecture.cs
Comment (MichalStrehovsky) [reply src/coreclr/tools/Common/TypeSystem/Common/TargetArchitecture.cs]: FWIW https://github.com/dotnet/runtime/pull/43706#discussion_r510754387 is the previous conversation on why we don't have TargetArchitecture.Armel.

**Comment by MichalStrehovsky** [comment]
> @MichalStrehovsky, I was motivated by aot related [improvements](https://github.com/dotnet/command-line-api/pulls?q=is%3Apr+author:adamsitnik+is:closed) made in System.CommandLine. The speed overhead is becoming insignificant.

So now it's about 3-4% slower E2E than with Internal.CommandLine. It's probably not blocking anymore. I would certainly not mind if it was faster though - Internal.CommandLine has similar ergonomics and feels like a fair benchmark.

For completeness, here's the size:

crossgen2 trimmed, compiled with R2R: 33,472,412 bytes -> 34,388,937 bytes (2.7% regression)
crossgen2 compiled with NativeAOT: 19,021,529 bytes -> 19,891,963 bytes (4.6% regression)

Not blocking either; could be better.

Cc @jonsequitur @adamsitnik

**Comment by jkotas** [comment]
1MB worth of code just to parse the command line the same way as before feels like a lot.

**Comment by am11** [comment]
> 1MB worth of code just to parse the command line the same way as before feels like a lot.

Isn't it 0.91 MB (R2R) and 0.87 MB (AoT)?

**Comment by jkotas** [comment]
Yes, I was rounding it. (Also, the existing lightweight parser is some code too.)

**Comment by am11** [comment]
Ah, I am sorry. I was reading it on a smaller screen and read it as `eleven` instead of `one` :sweat_smile:

I agree that we should continue to improve System.CommandLine (size, memory and speed), following the footsteps of @adamsitnik. :sunglasses:

File: src/coreclr/tools/aot/crossgen2/Properties/Resources.resx
Comment (jkotas) [inline src/coreclr/tools/aot/crossgen2/Properties/Resources.resx:394]: We got rid of the `Microsoft (R)` banner from .NET SDK tools. Could you please remove it from crossgen2 and ilc as well?

Comment (am11) [reply src/coreclr/tools/aot/crossgen2/Properties/Resources.resx]: Also removed from il{d}asm.

**Comment by jkotas** [comment]
Looks like that System.CommandLine exposed a bug in the runtime. `Build OSX x64 release Runtime_Debug` is crashing with:
```
 Emitting R2R PE file: /Users/runner/work/1/s/artifacts/obj/Microsoft.NETCore.App.Crossgen2/Release/net7.0/osx-x64/S.P.C.tmp
  
  Assert failure(PID 93502 [0x00016d3e], Thread: 279072 [0x44220]): SyncTableEntry::GetSyncTableEntry()[sbIndex].m_Object == obj
      File: /Users/runner/work/1/s/src/coreclr/vm/syncblk.cpp Line: 2083
      Image: /Users/runner/work/1/s/artifacts/bin/coreclr/OSX.x64.Release/crossgen2/osx-x64/publish/crossgen2
```

**Comment by am11** [comment]
Looks like an intermittent failure. I cannot reproduce it on osx-x64 with:
 `./build.sh -arch x64 -os OSX -c release -runtimeConfiguration debug  /p:CrossBuild=false`.

This leg was passing in penultimate commit, the last commit is only text update. Should we rerun the job?

**Comment by am11** [comment]
@jkotas, could you please also run the `runtime-dev-innerloop` pipeline. The failing OSX leg belongs to that pipeline.

**Comment by jkotas** [comment]
> 1MB worth of code just to parse the command line the same way as before feels like a lot.

@jonsequitur @adamsitnik Do you happen to have analysis where this is coming from?

I am hesitant to hit the merge button on this without understanding where this regression came from.

**Comment by MichalStrehovsky** [comment]
I have some data for the R2R version.

There is also 246 kB of satellite assemblies that I've not included in the breakdown below.

The size win in crossgen2.dll is basically the size of the current command line parser with ReadyToRun.

It would be good to know if System.CommandLine being 320 kB and depending on System.Net.Primitives, System.Runtime.Serialization.Formatters, System.ComponentModel, System.Private.Uri is expected.

Assembly | Before | After | Delta
-- | -- | -- | --
crossgen2.dll | 229,376 | 163,840 | -65,536
ILCompiler.DependencyAnalysisFramework.dll | 110,592 | 110,592 | 0
ILCompiler.Diagnostics.dll | 57,344 | 57,344 | 0
ILCompiler.ReadyToRun.dll | 1,581,056 | 1,581,056 | 0
ILCompiler.TypeSystem.dll | 573,440 | 573,440 | 0
System.Collections.Concurrent.dll | 94,208 | 94,208 | 0
System.Collections.dll | 94,208 | 94,208 | 0
System.Collections.Immutable.dll | 57,344 | 57,344 | 0
System.CommandLine.dll | | 327,680 | 327,680
System.ComponentModel.dll | | 20,480 | 20,480
System.Console.dll | 90,112 | 94,208 | 4,096
System.Diagnostics.FileVersionInfo.dll | 36,864 | 36,864 | 0
System.Diagnostics.StackTrace.dll | 28,672 | 28,672 | 0
System.Diagnostics.TraceSource.dll | 36,864 | 36,864 | 0
System.IO.Compression.dll | 225,280 | 225,280 | 0
System.IO.Compression.ZipFile.dll | 24,576 | 24,576 | 0
System.IO.MemoryMappedFiles.dll | 57,344 | 57,344 | 0
System.Linq.dll | 167,936 | 196,608 | 28,672
System.Memory.dll | 65,536 | 65,536 | 0
System.Net.Primitives.dll | | 57,344 | 57,344
System.Private.CoreLib.dll | 5,472,256 | 5,570,560 | 98,304
System.Private.Uri.dll | | 172,032 | 172,032
System.Private.Xml.dll | 557,056 | 557,056 | 0
System.Reflection.Metadata.dll | 651,264 | 651,264 | 0
System.Runtime.Serialization.Formatters.dll | | 28,672 | 28,672
System.Security.Cryptography.dll | 90,112 | 90,112 | 0
System.Text.Json.dll | 200,704 | 200,704 | 0
System.Threading.Tasks.Parallel.dll | 73,728 | 73,728 | 0
  | 10,575,872 | 11,247,616 | 671,744

**Comment by jonsequitur** [comment]
> It would be good to know if System.CommandLine being 320 kB and depending on System.Net.Primitives, System.Runtime.Serialization.Formatters, System.ComponentModel, System.Private.Uri is expected.

The optimization work so far has focused on startup time and speed rather than trimmed size, so this isn't entirely unexpected. It's certainly something we can take a look at reducing or relayering.

Off the top of my head, I'm guessing some of these dependencies are coming from types we bind by default. The heuristic and largely reflection- or `TypeConverter`-based parameter binding was replaced by a predefined set of binders for commonly-used types. This is why the dependencies currently exist on [System.Runtime.Serialization](https://github.com/dotnet/command-line-api/blob/4712b0cb160132a961bd165f59db8fe721df1162/src/System.CommandLine/Binding/ArgumentConverter.DefaultValues.cs#L8) and [System.Private.Uri](https://github.com/dotnet/command-line-api/blob/4dcee4d8a84a5aec7bbee0e9e112a2432e19a411/src/System.CommandLine/Binding/ArgumentConverter.StringConverters.cs#L302-L312), for example.

**Comment by jkotas** [comment]
> This is why the dependencies currently exist on [System.Runtime.Serialization](https://github.com/dotnet/command-line-api/blob/4712b0cb160132a961bd165f59db8fe721df1162/src/System.CommandLine/Binding/ArgumentConverter.DefaultValues.cs#L8) and [System.Private.Uri](https://github.com/dotnet/command-line-api/blob/4dcee4d8a84a5aec7bbee0e9e112a2432e19a411/src/System.CommandLine/Binding/ArgumentConverter.StringConverters.cs#L302-L312), for example.

Are there APIs in System.CommandLine that avoid the automatic binding to these types and pulling these types into dependency closure?

**Comment by jonsequitur** [comment]
Not currently, but it would be easy to move them into an opt-in method (that would be included in [`UseDefaults`](https://github.com/dotnet/command-line-api/blob/a1958c04615bc9def16f855009adf7f33bd067fd/src/System.CommandLine/Builder/CommandLineBuilderExtensions.cs#L318-L331) on the `CommandLineBuilder`). I suspect this approach should make this code eligible for trimming.

**Comment by jkotas** [comment]
> Not currently, but it would be easy to move them into an opt-in method (that would be included in [UseDefaults](https://github.com/dotnet/command-line-api/blob/a1958c04615bc9def16f855009adf7f33bd067fd/src/System.CommandLine/Builder/CommandLineBuilderExtensions.cs#L318-L331) on the CommandLineBuilder). I suspect this approach should make this code eligible for trimming.

I have opened https://github.com/dotnet/command-line-api/issues/1794 on this. Do you think you will get to it sometime soon?

**Comment by jkotas** [comment]
FWIW, it feels like something that would greatly benefit from source generators for the command line parser. The source generator would be in the best position to automatically reference the required convertors.

**Comment by am11** [comment]
> it feels like something that would greatly benefit from source generators for the command line parser.

cc @Keboo, who put forth the initial source generator work https://github.com/dotnet/command-line-api/pull/1362.

**Comment by jonsequitur** [comment]
> FWIW, it feels like something that would greatly benefit from source generators for the command line parser. The source generator would be in the best position to automatically reference the required convertors, and nothing more.

@KathleenDollard has a design and working code for a source generator-based layer on top of the core System.CommandLine APIs.

**Comment by jonsequitur** [comment]
> I have opened https://github.com/dotnet/command-line-api/issues/1794 on this. Do you think you will get to it sometime soon?

I'll take a look at it this week.

**Comment by jkotas** [comment]
@am11 Would you mind to submit the `Microsoft (R)` and the command line option rename as separate PR that can go into .NET 7?

I am leaning towards holding off on this change until the tree opens for .NET 8. There is risk of regressions from switching the command line parser (we do not have comprehensive command line tests for crossgen2). Also I hope that we will get the size regression down to more reasonable number by waiting for System.CommandLine updates.

**Comment by jkotas** [comment]
> There is also 246 kB of satellite assemblies that I've not included in the breakdown below.

@MichalStrehovsky Do you know what would need to be fixed to get this 246 kB of satellite assemblies trimmed? crossgen2 is not localized so it doesn't make sense for it to be carrying this.

File: src/coreclr/tools/aot/crossgen2/crossgen2.props
Comment (am11) [inline src/coreclr/tools/aot/crossgen2/crossgen2.props:29]: @jkotas, if we are not planning to add translations in Crossgen2 and ILC, shall we remove these embedded resources? I removed them from ilc. above (it was unused there). I can inline all `SR.` strings in Crossgen2.

Comment (jkotas) [reply src/coreclr/tools/aot/crossgen2/crossgen2.props:29]: We have flipped several times on dotnet/runtime tools localization. I would leave it as is. There is non-zero chance that we decide to do that in future.

**Comment by MichalStrehovsky** [comment]
> @MichalStrehovsky Do you know what would need to be fixed to get this 246 kB of satellite assemblies trimmed? crossgen2 is not localized so it doesn't make sense for it to be carrying this.

I don't know much about how satellite assemblies work - would it be sufficient to have a mode in illink that simply doesn't copy these over?

illink currently copies all of these over if a ResourceManager..ctor call is seen.

I've also filed https://github.com/dotnet/command-line-api/issues/1800 in the command line repo to make the satellite assemblies smaller. 5% of the size of these assemblies is a pompous assembly description that doesn't _need_ to be there.

**Comment by ViktorHofer** [comment]
> I don't know much about how satellite assemblies work - would it be sufficient to have a mode in illink that simply doesn't copy these over?

In case you want to avoid pulling in the extra satellite assemblies and just a default locale, you can set the following in the project file: `<SatelliteResourceLanguages>en-US</SatelliteResourceLanguages>`

https://docs.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props#satelliteresourcelanguages

**Comment by jkotas** [comment]
> In case you want to avoid pulling in the extra satellite assemblies and just a default locale, you can set the following in the project file: <SatelliteResourceLanguages>en-US</SatelliteResourceLanguages>

Nice! https://github.com/dotnet/runtime/pull/72478

**Comment by AntonLapounov** [comment]
> Looks like that System.CommandLine exposed a bug in the runtime. `Build OSX x64 release Runtime_Debug` is crashing with:
> 
> ```
>  Emitting R2R PE file: /Users/runner/work/1/s/artifacts/obj/Microsoft.NETCore.App.Crossgen2/Release/net7.0/osx-x64/S.P.C.tmp
>   
>   Assert failure(PID 93502 [0x00016d3e], Thread: 279072 [0x44220]): SyncTableEntry::GetSyncTableEntry()[sbIndex].m_Object == obj
>       File: /Users/runner/work/1/s/src/coreclr/vm/syncblk.cpp Line: 2083
>       Image: /Users/runner/work/1/s/artifacts/bin/coreclr/OSX.x64.Release/crossgen2/osx-x64/publish/crossgen2
> ```

I just hit the same assertion failure running crossgen2 with GCStress=3 on Windows ARM64 (without this PR).

**Comment by trylek** [comment]
@AntonLapounov - is the failure deterministic? I suspect it's worth investigating, I'm just wondering how hard it is to repro.

**Comment by AntonLapounov** [comment]
It took more than one day on my old ARM64 laptop to hit this.  I have a dump if anyone is interested.

**Comment by MichalStrehovsky** [comment]
One more thing I noticed is that if the root command has arguments (like in this case the list of input files), System.CommandLine will send all unmatched command line switches as the file names instead of erroring out. I don't know if there's a way not to do that.

So `crossgen2.exe input1.dll some_path\*.dll --SomeSwitchNameWithATypo` would treat `--SomeSwitchNameWithATypo` as a file name and we will crash with some nonsensical stack.

The Internal.CommandLine parser that we use right now will instead tell the user that `--SomeSwitchNameWithATypo` is an unknown switch.

**Comment by am11** [comment]
I think we shouldn't care about user experience of crossgen2.exe any more than dotnet.exe (that uses System.CommandLine and has much vaster public usage).

Of course if we can configure the behavior of the open ended args, that would be nice.

To wit:
```sh
$ ~/.dotnet8/dotnet --foo
The command could not be loaded, possibly because:
  * You intended to execute a .NET application:
      The application '--foo' does not exist.
  * You intended to execute a .NET SDK command:
      No .NET SDKs were found.

```

**Comment by am11** [comment]
Resolved merge conflicts and unified `--parallelism` defaults (ported https://github.com/dotnet/runtime/pull/74956/files#diff-58625e1414cc2e244a276695b0572cc258c7ac4697f8e4326614bd2745fc27aa to ilc).

@jkotas, tree is opened for .NET 8, should we get this in?

**Comment by jkotas** [comment]
There are ongoing discussions about the future System.CommandLine. I would prefer for to wait for those discussions to come to conclusion.

**Comment by am11** [comment]
ILVerify, R2Rdump etc. are depending on the same SystemCommandLine package. Getting this in will make it easier to upgrade all dependents to System.CommandLine's future release, when it's time to update the package.

tbh, merge conflicts in this PR are not fun to resolve.

**Comment by jonsequitur** [comment]
> One more thing I noticed is that if the root command has arguments (like in this case the list of input files), System.CommandLine will send all unmatched command line switches as the file names instead of erroring out. I don't know if there's a way not to do that.

It should be able to handle this, @MichalStrehovsky. I couldn't reproduce the issue. Configuring more specific expected types and adding validators might help.

If the parser is throwing an exception, that's unambiguously a bug. Do you have a stack trace for this issue?

**Comment by jkotas** [comment]
> It should be able to handle this, @MichalStrehovsky. I couldn't reproduce the issue.

```
dotnet tool install dotnet-ilverify --global --version 7.0.0-rc.1.22426.10`
ilverify --referenci c:\Windows\Microsoft.NET\Framework\v4.0.30319\mscorlib.dll
```

It prints `Error: No files matching --referenci`. It would be better if it printed "Error: Bad option --referenci`.

**Comment by jkotas** [comment]
@jonsequitur I have opened https://github.com/dotnet/command-line-api/issues/1860 on this

**Comment by jkotas** [comment]
> Getting this in will make it easier to upgrade all dependents to System.CommandLine's future release, when it's time to update the package.

It is not necessary correct. It can be more work to update the package if there are breaking changes.

I agree with that there are number of tools depending on System.CommandLine, and having two more is not going to change the update problem much. So let's get this merged. There is a lot of runway in .NET 8 to the issues sorted out.

**Review by jkotas — APPROVED** [review]
Thanks!

**Comment by am11** [comment]
@jkotas, thank you for your help and support. I will keep an eye out on https://github.com/dotnet/runtime/issues/68578 and will follow up once the new version is released.

I am working on dotnet-pgo to consume System.CommandLine, so we can delete the internal CommandLine implementation.

**Pull request merged**
